src/python/pants/backend/python/util_rules/pex_cli.py

```diff
@@ -130,6 +204,114 @@ async def download_pex_pex(pex_cli: PexCli, platform: Platform) -> PexPEX:
+    # where Pants rule code cannot directly control where the credentials are stored.
+    if not env_tgt.can_access_local_system_paths:
+        return _KeyringState(keyring_data_path=None)
+
+    keyring_plugin_request_types = union_membership.get(PexKeyringConfigurationRequest)
+    if not keyring_plugin_request_types:
+        return _KeyringState(keyring_data_path=None)
+
+    credentials_responses: dict[str, list[tuple[str, tuple[str, str]]]] = {}
+    for keyring_plugin_request_type in keyring_plugin_request_types:
+        keyring_plugin_request = keyring_plugin_request_type()
+        keyring_plugin_response = await Get(  # noqa: PNT30: Only one provider expected.
+            PexKeyringConfigurationResponse, PexKeyringConfigurationRequest, keyring_plugin_request
+        )
+
+        keyring_plugin_credentials = keyring_plugin_response.credentials
+        if not keyring_plugin_credentials:
+            continue
+
+        for site, user_password in keyring_plugin_credentials.items():
+            if site not in credentials_responses:
+                credentials_responses[site] = []
+            credentials_responses[site].append((keyring_plugin_request_type.name, user_password))
+
+    # Short circuit if the providers did not supply any credentials.
+    if not credentials_responses:
+        return _KeyringState(keyring_data_path=None)
+
+    # Check for multiple responses for a single site. Keyring and our simulation of it only supports
+    # a single user/password per site.
+    credentials: list[tuple[str, str, str]] = []
+    for site, user_password_creds_by_provider in credentials_responses.items():
+        if len(user_password_creds_by_provider) > 1:
+            providers = [x[0] for x in user_password_creds_by_provider]
+            raise Exception(
+                f"Multiple keyring plugins returned responses for `{site}`. "
+                "The keyring support only supports a single user/password per site. "
+                f"The keyring plugins in conflict are: {', '.join(providers)}."
+            )
+        user_password_creds = user_password_creds_by_provider[0][1]
+        credentials.append((site, user_password_creds[0], user_password_creds[1]))
+    credentials.sort()
+
+    # Write the credentials to a file based on the current run ID.
+    content = StringIO()
+    content.write("declare -A pants_keyring_credentials\n")
+    for site, user, password in credentials:
+        content.write(
+            f"pants_keyring_credentials[{_always_shlex_quote(site)}]={_always_shlex_quote(user)}\n"
+        )
+        site_user = f"{site}:{user}"
+        content.write(
+            f"pants_keyring_credentials[{_always_shlex_quote(site_user)}]={_always_shlex_quote(password)}\n"
+        )
+
+    # TODO: Consider removing the run ID so that this value does not contribute to the cache key
+    # and cause unneeded invalidation of build actions as it changes for each run.
```

Comment:
This is probably a main issue of this PR. Where do we store the credentials out of band in a way that does not cause unnecessary invalidation of processes across different Pants runs?

Reply:
The concern is that the credentials file gets rewritten on concurrent Pants runs. Maybe it should be stored per-pantsd process?

Reply:
One alternative solution might install the `keyring` package somewhere (whether manually by the user or Pants doing it) and then set the password by calling `keyring set SITE USER PASSWORD`. Pants would no longer use a shim binary in that case.

An advantages is that we would no longer need an environment variable to point at the data file. This would prevent invalidating for changes to that filename (e.g., for different run/session IDs).

It assumes that updating the password is idempotent in the sense that any caller of `keyring get SITE USER` will be able to use that password. This should in fact be the case given how keyring works.

Reply:
Can pants not depend on `keyring` and use it programmatically in-process?

Reply:
Perhaps I am misunderstanding, but IIUC `pip` reads from the system keyring. So couldn't Pants use `keyring` to write creds to it first?

Reply:
> but IIUC pip reads from the system keyring. 

That assumes the user has installed the `keyring` PyPi package globally.

Reply:
Since it is hard to setup the in-process "import" `keyring` in pip with Pex given hermeticity concerns. See John's release notes for Pex v2.29.0 at https://github.com/pex-tool/pex/blob/main/CHANGES.md#2290.

Reply:
So the trade off made was, we cannot assume the in-process support works and we cannot assume the user has installed a global `keyring` binary, so with this PR, Pants will try to inject a `keyring` binary for Pip to see and Pip will need to be in `--keyring-provider=subprocess`.

Reply:
That said, maybe the trade off of requiring an installed `keyring` provider will actually be the better solution?

Reply:
If that is the case, then this PR would reduce to just a `--pex-cli-keyring-provider` option instead of a plugin API? The PR stacked on top of this one to add AWS CodeArtifact token renewal is the driving force for this PR.

What are your thoughts on such token auto-renewal being driven from within a Pants plugin (which can be invoked on a per-session basis)?

An alternative solution would give users the ability to invoke a runnable during session startup as configured by a new target type. Then the user could define the aws token logic in a BUiLD file. Then there would be no need for most of the current solution direction.

Reply:
I think the simplest possible solution, that punts installing, locking down and configuring a keychain binary to the user, is fine...